dissect/target/plugins/os/unix/log/journal.py

```diff
@@ -0,0 +1,389 @@
+};
+
+struct Header {
+    uint8_t     signature[8];
+    le32_t      compatible_flags;
+    le32_t      incompatible_flags;
+    State       state;
+    uint8_t     reserved[7];
+    sd_id128_t  file_id;
+    sd_id128_t  machine_id;
+    sd_id128_t  tail_entry_boot_id;
+    sd_id128_t  seqnum_id;
+    le64_t      header_size;
+    le64_t      arena_size;
+    le64_t      data_hash_table_offset;
+    le64_t      data_hash_table_size;
+    le64_t      field_hash_table_offset;
+    le64_t      field_hash_table_size;
+    le64_t      tail_object_offset;
+    le64_t      n_objects;
+    le64_t      n_entries;
+    le64_t      tail_entry_seqnum;
+    le64_t      head_entry_seqnum;
+    le64_t      entry_array_offset;
+    le64_t      head_entry_realtime;
+    le64_t      tail_entry_realtime;
+    le64_t      tail_entry_monotonic;
+    le64_t      n_data;
+    le64_t      n_fields;
+    le64_t      n_tags;
+    le64_t      n_entry_arrays;
+    le64_t      data_hash_chain_depth;
+    le64_t      field_hash_chain_depth;
+    le32_t      tail_entry_array_offset;
+    le32_t      tail_entry_array_n_entries;
+    le64_t      tail_entry_offset;
+};
+
+enum ObjectType : uint8 {
+    OBJECT_UNUSED,
+    OBJECT_DATA,
+    OBJECT_FIELD,
+    OBJECT_ENTRY,
+    OBJECT_DATA_HASH_TABLE,
+    OBJECT_FIELD_HASH_TABLE,
+    OBJECT_ENTRY_ARRAY,
+    OBJECT_TAG,
+    _OBJECT_TYPE_MAX
+};
+
+enum ObjectFlag : uint8 {
+    OBJECT_UNCOMPRESSED      =  0,
+    OBJECT_COMPRESSED_XZ     =  1,
+    OBJECT_COMPRESSED_LZ4    =  2,
+    OBJECT_COMPRESSED_ZSTD   =  4,
+    _OBJECT_COMPRESSED_MASK  =  7
+};
+
+struct ObjectHeader {
+    ObjectType  type;               // The type field is one of the object types listed above
+    uint8_t     flags;              // If DATA object the value is ObjectFlag
+    uint8_t     reserved[6];
+    le64_t      size;               // The size field encodes the size of the object including all its headers and payload
+};
+
+
+// The first four members are copied from ObjectHeader, so that the size can be used as the length of payload
+struct DataObject {
+    ObjectType  type;
+    ObjectFlag  flags;
+    uint8_t     reserved[6];
+    le64_t      size;
+    le64_t      hash;
+    le64_t      next_hash_offset;
+    le64_t      next_field_offset;
+    le64_t      entry_offset;
+    le64_t      entry_array_offset;
+    le64_t      n_entries;
+    char        payload[size - 64];  // Data objects carry actual field data in the payload[] array.
+};
+
+struct EntryItem {
+    le32_t object_offset;
+    le32_t unknown1;
+    le64_t unknown2;
+};
+
+// The first four members are copied from ObjectHeader, so that the size can be used as the length of items
+struct EntryObject {
+    ObjectType  type;
+    uint8_t     flags;
+    uint8_t     reserved[6];
+    le64_t      size;
+    le64_t      seqnum;
+    le64_t      realtime;
+    le64_t      monotonic;
+    sd_id128_t  boot_id;
+    le64_t      xor_hash;
+    EntryItem   items[size / 64];
```

Comment:
Should this not be 32 or 64 depending on HEADER_INCOMPATIBLE_COMPACT?

Reply:
Is adding additional c-struct the preferred way of doing this?

Reply:
Yes, I would use double c-structures. That looks like the easiest way to go.

I think something like:

```
 if self.header.incompatible_flags == c_journal.IncompatibleFlag.HEADER_INCOMPATIBLE_COMPACT:
   entry_array_object = c_journal.EntryArrayObject_Compact(self.fh)
else:
   entry_array_object = c_journal.EntryArrayObject(self.fh)
```